src/frontend/src/optimizer/plan_visitor/cardinality_visitor.rs

```diff
@@ -0,0 +1,425 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::HashSet;
+use std::ops::{Mul, Sub};
+
+use risingwave_pb::plan_common::JoinType;
+
+use self::card::Cardinality;
+use super::{DefaultBehavior, DefaultValue, PlanVisitor};
+use crate::catalog::system_catalog::pg_catalog::PG_NAMESPACE_TABLE_NAME;
+use crate::optimizer::plan_node::generic::Limit;
+use crate::optimizer::plan_node::{self, PlanTreeNode, PlanTreeNodeBinary, PlanTreeNodeUnary};
+use crate::optimizer::plan_visitor::PlanRef;
+
+pub mod card {
+    use std::cmp::{max, min, Ordering};
+    use std::ops::{Add, Mul, Sub};
+
+    /// The upper bound of the [`Cardinality`].
+    #[derive(Clone, Copy, Debug, PartialEq, Eq)]
+    pub enum Hi {
+        Limited(usize),
+        Unlimited,
+    }
+
+    impl PartialOrd for Hi {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+
+    impl Ord for Hi {
+        fn cmp(&self, other: &Self) -> Ordering {
+            match (self, other) {
+                (Self::Unlimited, Self::Unlimited) => Ordering::Equal,
+                (Self::Unlimited, Self::Limited(_)) => Ordering::Greater,
+                (Self::Limited(_), Self::Unlimited) => Ordering::Less,
+                (Self::Limited(lhs), Self::Limited(rhs)) => lhs.cmp(rhs),
+            }
+        }
+    }
+
+    impl From<Option<usize>> for Hi {
+        fn from(value: Option<usize>) -> Self {
+            value.map_or(Self::Unlimited, Self::Limited)
+        }
+    }
+
+    impl From<usize> for Hi {
+        fn from(value: usize) -> Self {
+            Self::Limited(value)
+        }
+    }
+
+    /// The cardinality of the output rows of a plan node. Bounds are inclusive.
+    ///
+    /// The default value is `0..`, i.e. the number of rows is unknown.
+    // TODO: Make this the property of each plan node.
+    #[derive(Clone, Copy, Debug)]
+    pub struct Cardinality {
+        lo: usize,
+        hi: Hi,
+    }
+
+    impl Default for Cardinality {
+        fn default() -> Self {
+            Self {
+                lo: 0,
+                hi: None.into(),
+            }
+        }
+    }
+
+    impl<T> From<T> for Cardinality
+    where
+        T: std::ops::RangeBounds<usize>,
+    {
+        fn from(value: T) -> Self {
+            let lo = match value.start_bound() {
+                std::ops::Bound::Included(lo) => *lo,
+                std::ops::Bound::Excluded(lo) => lo.saturating_add(1),
+                std::ops::Bound::Unbounded => 0,
+            };
+            let hi = match value.end_bound() {
+                std::ops::Bound::Included(hi) => Hi::Limited(*hi),
+                std::ops::Bound::Excluded(hi) => Hi::Limited(hi.saturating_sub(1)),
+                std::ops::Bound::Unbounded => Hi::Unlimited,
+            };
+            Self::new(lo, hi)
+        }
+    }
+
+    impl Cardinality {
```

Comment:
Is it better to remove `exact`, `at_least`, `limit_to`, and `as_low_as`? When I read the code, I will keep translating these abstracts into range bounds. Using range bounds directly seems more straightforward.

Reply:
Sure 🥵